### PR TITLE
src/signature: get key passphrase from environment

### PR DESCRIPTION
**Pull request by marckleinebudde**
The user can add the passphrase of the private key to the environment
variable RAUC_KEY_PASSPHRASE. Pass this to PEM_read_bio_PrivateKey()
if set.

Signed-off-by: Marc Kleine-Budde <mkl@pengutronix.de>

### CODE REVIEW

File: docs/advanced.rst [hunk 1]
```diff
@@ -150,6 +150,18 @@ To do this, specify the ``--intermediate`` argument during bundle creation::
 Note that you can specify the ``--intermediate`` argument multiple times to
 include multiple intermediate certificates to your bundle signature.
 
+Passphrase Handling
+~~~~~~~~~~~~~~~~~~~
+
+If the signing key is protected by a passphrase it has to entered
+during signing operation. For automation purpose the passphrase can be
+set in the environment variable RAUC_KEY_PASSPHRASE.
```
Comment (jluebbe) [inline docs/advanced.rst:158]:
```suggestion
If the signing key is protected by a passphrase it has to entered
during signing operations.
For automation purposes the passphrase can be
set using the environment variable RAUC_KEY_PASSPHRASE.
```

File: docs/advanced.rst [hunk 2]
```diff
@@ -150,6 +150,18 @@ To do this, specify the ``--intermediate`` argument during bundle creation::
...
+
+.. note::
+  Since the environment of other processes is visible on certain
+  platforms (e.g. ps under certain Unix OSes) this option should be
+  used with caution.
```
Comment (jluebbe) [inline docs/advanced.rst:163]: We only support Linux.

File: docs/reference.rst [hunk 3]
```diff
@@ -925,6 +925,7 @@ Command Line Tool
     write-slot            Write image to slot and bypass all update logic
 
   Environment variables:
+    RAUC_KEY_PASSPHRASE Passphrase to use for accessing keys files (signing only)
```
Comment (jluebbe) [inline docs/reference.rst:928]:
```suggestion
    RAUC_KEY_PASSPHRASE Passphrase to use for accessing key files (signing only)
```

File: rauc.1 [hunk 4]
```diff
@@ -328,6 +328,10 @@ Write image to slot and bypass all update logic.
 
 .SH ENVIRONMENT
 
+.TP
+.B RAUC_KEY_PASSPHRASE
+Passphrase to use for accessing keys files (signing only)
```
Comment (jluebbe) [inline rauc.1:333]:
```suggestion
Passphrase to use for accessing key files (signing only)
```

File: src/signature.c [hunk 5]
```diff
@@ -188,6 +188,7 @@ static EVP_PKEY *load_key_file(const gchar *keyfile, GError **error)
 	BIO *key = NULL;
 	unsigned long err;
 	const gchar *data;
+	gchar *passphrase;
```
Comment (jluebbe) [inline src/signature.c:191]:
```suggestion
	const gchar *passphrase;
```

**Review by jluebbe — CHANGES_REQUESTED** [review]
We should probably add a test case as well.

File: docs/advanced.rst (hunk 2 above)
Comment (marckleinebudde) [reply docs/advanced.rst]: That was copied from the openssl man page :) fixed

File: src/signature.c (hunk 5 above)
Comment (marckleinebudde) [reply src/signature.c:191]:
```
src/signature.c:210:56: warning: passing argument 4 of 'PEM_read_bio_PrivateKey' discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
  210 |         res = PEM_read_bio_PrivateKey(key, NULL, NULL, passphrase);
      |                                                        ^~~~~~~~~~
```

Comment (marckleinebudde) [reply src/signature.c:191]: Too many deprecation warnings with openssl-3 :crying_cat_face:
I've taken the `g_strdup()` + `g_free()` route...

File: docs/reference.rst (hunk 3 above)
Comment (marckleinebudde) [reply docs/reference.rst]: fixed

File: rauc.1 (hunk 4 above)
Comment (marckleinebudde) [reply rauc.1]: fixed

File: src/signature.c (hunk 5 above)
Comment (jluebbe) [reply src/signature.c]: In cases like these (where we know that OpenSSL won't write to this pointer), we've casted away const in the function call.

File: docs/advanced.rst [hunk 6]
```diff
@@ -150,6 +150,18 @@ To do this, specify the ``--intermediate`` argument during bundle creation::
 Note that you can specify the ``--intermediate`` argument multiple times to
 include multiple intermediate certificates to your bundle signature.
 
+Passphrase Handling
+~~~~~~~~~~~~~~~~~~~
+
+If the signing key is protected by a passphrase it has to entered
```
Comment (ejoerns) [inline docs/advanced.rst:156]:
```suggestion
If the signing key is protected by a passphrase it has to be entered
```

File: docs/advanced.rst [hunk 7]
```diff
@@ -150,6 +150,18 @@ To do this, specify the ``--intermediate`` argument during bundle creation::
...
+during signing operations.
+For automation purposes the passphrase can be set using the
+environment variable RAUC_KEY_PASSPHRASE.
```
Comment (ejoerns) [inline docs/advanced.rst:159]:
```suggestion
environment variable ``RAUC_KEY_PASSPHRASE``.
```

File: docs/advanced.rst [hunk 8]
```diff
@@ -150,6 +150,18 @@ To do this, specify the ``--intermediate`` argument during bundle creation::
...
+
+.. note::
+  Since the environment of other processes is visible on Linux this
```
Comment (ejoerns) [inline docs/advanced.rst:162]:
```suggestion
  Since the environment of other processes is visible on Linux, this
```

**Review by ejoerns — CHANGES_REQUESTED** [review]

**Review by ejoerns — APPROVED** [review]

**Review by jluebbe — APPROVED** [review]

**Comment by jluebbe** [comment]
Thanks for your contribution!

**Pull request merged**